### PR TITLE
Fix MISRA 11.6 violations

### PR DESCRIPTION
**Pull request by alfred2g**
Fix MISRA 11.6 violations

Description
-----------
Fix MISRA 11.6 violations

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/FreeRTOS_IP_Utils.c [hunk 1]
```diff
@@ -825,6 +829,10 @@ uint16_t usGenerateChecksum( uint16_t usSum,
     xTerm.u32 = 0UL;
 
     xSource.u8ptr = ipPOINTER_CAST( uint8_t *, pucNextData );
+
+    /* the conversion here does not cause a loss of data, as uintptr_t fits a
+     * pointer type on the system */
+    /* coverity[misra_c_2012_rule_11_1_violation] */
```
Comment (paulbartell) [inline source/FreeRTOS_IP_Utils.c:835]: MISRA 11.1 is in regards to function pointers. Shouldn't this be referencing 11.4?

File: source/FreeRTOS_IP_Utils.c [hunk 2]
```diff
@@ -220,6 +220,10 @@ static NetworkBufferDescriptor_t * prvPacketBuffer_to_NetworkBuffer( const void
     else
     {
         /* Obtain the network buffer from the zero copy pointer. */
+
+        /* the conversion here does not cause a loss of data, as uintptr_t fits a
+         * pointer type on the system */
+        /* coverity[misra_c_2012_rule_11_1_violation] */
         uxBuffer = ipPOINTER_CAST( uintptr_t, pvBuffer );
```
Comment (paulbartell) [inline source/FreeRTOS_IP_Utils.c:227]: Something like:
```
uint8_t * pcBufferOffset = &( ( ( uint8_t * ) pvBuffer )[ 0 -uxOffset - ipBUFFER_PADDING ] );
```

Would also suffice.

Comment (htibosch) [reply source/FreeRTOS_IP_Utils.c:227]: About the following:
~~~
uint8_t * pcBufferOffset = &( ( ( uint8_t * ) pvBuffer )[ 0 - uxOffset - ipBUFFER_PADDING ] );
~~~
This is the transformation to the beginning of the allocated buffer.

In my experience, coverity want to see such a transformation in smaller steps:

uint8_t * pcBufferOffset = ( uint8_t * ) pvBuffer;
pcBufferOffset = &( pcBufferOffset [ 0 - ( uxOffset + ipBUFFER_PADDING ) ] );

or even more steps.

**Review by htibosch — APPROVED** [review]
Thank you for this PR.

File: source/FreeRTOS_Sockets.c [hunk 3]
```diff
@@ -1649,6 +1649,8 @@ void * vSocketClose( FreeRTOS_Socket_t * pxSocket )
                     ( ( pxOtherSocket->u.xTCP.bits.bPassQueued != pdFALSE_UNSIGNED ) ||
                       ( pxOtherSocket->u.xTCP.bits.bPassAccept != pdFALSE_UNSIGNED ) ) )
                 {
+                    /* recursive call */
+                    /* coverity[misra_c_2012_rule_17_2_violation] */
                     ( void ) vSocketClose( pxOtherSocket );
```
[the remark this answers is not in the document]
Comment (alfred2g) [reply source/FreeRTOS_Sockets.c:1654]: yes for sure, i'll add that
do we have a number on that limitation ?

File: source/FreeRTOS_IP_Utils.c (hunk 2 above)
Comment (shubnil) [reply source/FreeRTOS_IP_Utils.c:227]: Agree. Similar to the change https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/491/files

File: source/FreeRTOS_IP_Utils.c [hunk 4]
```diff
@@ -220,6 +220,10 @@ static NetworkBufferDescriptor_t * prvPacketBuffer_to_NetworkBuffer( const void
     else
     {
         /* Obtain the network buffer from the zero copy pointer. */
+
+        /* the conversion here does not cause a loss of data, as uintptr_t fits a
+         * pointer type on the system */
+        /* coverity[misra_c_2012_rule_11_1_violation] */
```
Comment (paulbartell) [inline source/FreeRTOS_IP_Utils.c:226]:
```suggestion
        /* coverity[misra_c_2012_rule_11_4_violation] */
```

**Review by AniruddhaKanhere — APPROVED** [review]

File: source/FreeRTOS_IP.c [hunk 5]
```diff
@@ -386,6 +386,7 @@ static void prvProcessIPEventsAndTimers( void )
                     eDHCPState_t eState;
 
                     /* Cast in two steps to please MISRA. */
```
Comment (paulbartell) [inline source/FreeRTOS_IP.c:388]: Needs a justification.

File: source/FreeRTOS_IP_Utils.c (hunk 2 above)
Comment (paulbartell) [reply source/FreeRTOS_IP_Utils.c]: Rule 11.1 refers to function pointers, but this is not a function pointer.

**Review by shubnil — APPROVED** [review]

**Review by shubnil — DISMISSED** [review]

**Review by paulbartell — APPROVED** [review]

**Review by paulbartell — APPROVED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by paulbartell — APPROVED** [review]

**Pull request merged**